runtime/interpreter/interpreter_expression.go

```diff
@@ -0,0 +1,669 @@
+	return BoolValue(expression.Value)
+}
+
+func (interpreter *Interpreter) VisitNilExpression(_ *ast.NilExpression) ast.Repr {
+	return NilValue{}
+}
+
+func (interpreter *Interpreter) VisitIntegerExpression(expression *ast.IntegerExpression) ast.Repr {
+	return IntValue{expression.Value}
+}
+
+func (interpreter *Interpreter) VisitFixedPointExpression(expression *ast.FixedPointExpression) ast.Repr {
+	// TODO: adjust once/if we support more fixed point types
+
+	value := fixedpoint.ConvertToFixedPointBigInt(
+		expression.Negative,
+		expression.UnsignedInteger,
+		expression.Fractional,
+		expression.Scale,
+		sema.Fix64Scale,
+	)
+
+	if expression.Negative {
+		return Fix64Value(value.Int64())
+	} else {
+		return UFix64Value(value.Uint64())
+	}
+}
+
+func (interpreter *Interpreter) VisitStringExpression(expression *ast.StringExpression) ast.Repr {
+	return NewStringValue(expression.Value)
+}
+
+func (interpreter *Interpreter) VisitArrayExpression(expression *ast.ArrayExpression) ast.Repr {
+	values := interpreter.visitExpressionsNonCopying(expression.Values)
+
+	argumentTypes := interpreter.Program.Elaboration.ArrayExpressionArgumentTypes[expression]
+	elementType := interpreter.Program.Elaboration.ArrayExpressionElementType[expression]
+
+	copies := make([]Value, len(values))
+	for i, argument := range values {
+		argumentType := argumentTypes[i]
+		copies[i] = interpreter.copyAndConvert(argument, argumentType, elementType)
+	}
+
+	return NewArrayValueUnownedNonCopying(copies...)
+}
+
+func (interpreter *Interpreter) VisitDictionaryExpression(expression *ast.DictionaryExpression) ast.Repr {
+	values := interpreter.visitEntries(expression.Entries)
+
+	entryTypes := interpreter.Program.Elaboration.DictionaryExpressionEntryTypes[expression]
+	dictionaryType := interpreter.Program.Elaboration.DictionaryExpressionType[expression]
+
+	dictionary := NewDictionaryValueUnownedNonCopying()
+	for i, dictionaryEntryValues := range values {
+		entryType := entryTypes[i]
+
+		key := interpreter.copyAndConvert(
+			dictionaryEntryValues.Key,
+			entryType.KeyType,
+			dictionaryType.KeyType,
+		)
+
+		value := interpreter.copyAndConvert(
+			dictionaryEntryValues.Value,
+			entryType.ValueType,
+			dictionaryType.ValueType,
+		)
+
+		// TODO: panic for duplicate keys?
+
+		// NOTE: important to convert in optional, as assignment to dictionary
+		// is always considered as an optional
+
+		locationRange := interpreter.locationRange(expression)
+		_ = dictionary.Insert(interpreter, locationRange, key, value)
+	}
+
+	return dictionary
+}
+
+func (interpreter *Interpreter) VisitMemberExpression(expression *ast.MemberExpression) ast.Repr {
+	result := interpreter.evalExpression(expression.Expression)
+	if expression.Optional {
+		switch typedResult := result.(type) {
+		case NilValue:
+			return typedResult
+
+		case *SomeValue:
+			result = typedResult.Value
+
+		default:
+			panic(errors.NewUnreachableError())
+		}
+	}
+
+	value := result.(Value)
```

Comment:
can omit the cast. `result` is already a `Value`

Reply:
Nice!

Reply:
Cleaned up in 9663ee9947b549463d4817a888cfa595c8180b63